File: autosubliminal/util/websocket.py
# coding=utf-8

from typing import Any, Dict, Literal

import autosubliminal

SYSTEM_START = 'SYSTEM_START'
SYSTEM_RESTART = 'SYSTEM_RESTART'
SYSTEM_SHUTDOWN = 'SYSTEM_SHUTDOWN'
SYSTEM_UPDATE = 'SYSTEM_UPDATE'
SCHEDULER_START = 'SCHEDULER_START'
SCHEDULER_FINISH = 'SCHEDULER_FINISH'
WANTED_ITEM_UPDATE = 'WANTED_ITEM_UPDATE'
WANTED_ITEM_DELETE = 'WANTED_ITEM_DELETE'

SUPPORTED_EVENT_TYPES = [
    SYSTEM_START,
    SYSTEM_RESTART,
    SYSTEM_SHUTDOWN,
    SYSTEM_UPDATE,
    SCHEDULER_START,
    SCHEDULER_FINISH,
    WANTED_ITEM_UPDATE,
    WANTED_ITEM_DELETE,
]

MessageSeverity = Literal['info', 'success', 'warn', 'error']


def send_websocket_event(type: str, data: Dict[str, Any] = None) -> None:
    """Send a websocket event message.

    :param type: the event type
    :type type: str
    :param data: the optional data for the event
    :type data: dict
    """
    event = {'type': 'EVENT', 'event': {'type': type, 'data': data}}
    autosubliminal.WEBSOCKETMESSAGEQUEUE.append(event)


def send_websocket_notification(message: str, severity: 'MessageSeverity' = 'info', sticky: bool = False) -> None:
    """Send a websocket notification message.

    :param message: the notification message
    :type message: str
    :param severity: the notification message severity
    :type severity: str
    :param sticky: indication if it will be shown at a fixed inline location and it will not fade
    :type sticky: bool
    """
    notification = {
        'type': 'NOTIFICATION',
        'notification': {'message': message, 'severity': severity, 'sticky': sticky},
    }
    autosubliminal.WEBSOCKETMESSAGEQUEUE.append(notification)

File: web/autosubliminal/src/app/shared/models/websocket.ts

// System websocket event received from server

import { MessageSeverity } from './message';
import { Scheduler } from './scheduler';
import { SystemUpdate } from './systemupdate';

export enum SystemWebSocketServerEventType {
  SYSTEM_START = 'SYSTEM_START',
  SYSTEM_RESTART = 'SYSTEM_RESTART',
  SYSTEM_SHUTDOWN = 'SYSTEM_SHUTDOWN',
  SYSTEM_UPDATE = 'SYSTEM_UPDATE',
  SCHEDULER_START = 'SCHEDULER_START',
  SCHEDULER_FINISH = 'SCHEDULER_FINISH',
  WANTED_ITEM_UPDATE = 'WANTED_ITEM_UPDATE',
  WANTED_ITEM_DELETE = 'WANTED_ITEM_DELETE'
}

export type SystemWebSocketServerEventData = Scheduler | SystemUpdate;

export class SystemWebSocketServerEvent {
  type = 'EVENT';
  event: {
    type: SystemWebSocketServerEventType;
    data: SystemWebSocketServerEventData;
  };

  constructor(obj: any) {
    Object.assign(this, obj);
  }
}

// System websocket notification received from server

export class SystemWebSocketServerNotification {
  type = 'NOTIFICATION';
  notification: {
    severity: MessageSeverity;
    sticky: boolean;
    message: string;
  };

  constructor(obj: any) {
    Object.assign(this, obj);
  }
}

// Websocket event sent from client
export enum SystemWebSocketClientEventType {
  RUN_SCHEDULER = 'RUN_SCHEDULER',
  RUN_SYSTEM_PROCESS = 'RUN_SYSTEM_PROCESS'
}

export class SystemWebSocketClientEvent {
  type = 'EVENT';
  event: {
    type: SystemWebSocketClientEventType;
    data: {
      name: string;
    }
  };

  constructor(type: SystemWebSocketClientEventType, name: string) {
    this.event = {
      type,
      data: {
        name
      }
    };
  }
}

export type SystemWebSocketServerMessage = SystemWebSocketServerEvent | SystemWebSocketServerNotification;

export type SystemWebSocketClientMessage = SystemWebSocketClientEvent;

export type SystemWebSocketMessage = SystemWebSocketServerMessage | SystemWebSocketClientMessage;

export type LogWebSocketMessage = string;
